text/2024-01-27-autohinting.md

```diff
@@ -0,0 +1,266 @@
+  surpasses the threshold.
+
+* Edges
+  * BC, FGH
+  * DE
+  * IJK, NA
+  * LM
+
+> Note that edges can be formed from non-contiguous segments.
+
+* Edge links (stems)
+  * BC, FGH <-> IJK, NA
+
+## Time for a change
+
+Once we've computed the topological features of our outline, we can start
+making adjustments. In general, grid-fitting is applied directly to our
+computed _edges_ and then those changes are propagated to the outline
+_points_.
+
+### Hinting edges ([src](https://github.com/googlefonts/fontations/blob/b283d608f7beca834c7d56cb649c79ba40c12411/skrifa/src/outline/autohint/hint/edges.rs))
+
+Edge hinting is performed in three passes:
+
+1. Edges that have been assigned a blue zone are handled first. These edges are
+   simply snapped to the grid-fitted position of their associated blue zone.
+2. Stem edges (those that have been linked with another edge to form a pair) are
+   handled next. These are snapped directly to the grid without guidance from an
+   alignment zone. The hinter tries to maintain a minimal distance from already
+   hinted edges to avoid overlaps.
+3. Remaining edges (single point edges and serifs) are handled last. These don't
+   form stems so are just fitted independently.
+
+The edges processed by the first two passes are always composed of pairs. In these
+cases, the first edge is snapped to either an alignment zone or the closest pixel.
+The position of the second edge is computed by mapping the distance between the two
+edges to the closest matching "standard stem width" from the set described in the
+metrics section above.
+
+### Hinting points ([src](https://github.com/googlefonts/fontations/blob/main/skrifa/src/outline/autohint/hint/outline.rs))
+
+With grid-fitted edges available, the final step is to apply these adjustments to the
+points of the original outline.
+
+Points are adjusted in three groups:
+
+1. Points that directly contributed to edge formation are adjusted first. We
+   iterate through the segments that make up each edge and then through the
+   points that make up each segment and apply the adjustment. For "strong"
+   hinting, all points are snapped directly to the position of the edge. In
+   lighter hinting modes, points are shifted by the difference between the
+   original and grid-fitted edge position.
+2. Points that weren't touched in the previous pass and are also not flagged
+   for weak interpolation are adjusted next. For these points, we find the
+   nearest enclosing edges and perform a linear interpolation based on the
+   adjustments made to those two edges.
+3. Finally, all remaining points are adjusted using the exactly same procedure
+   as IUP (interpolate untouched points) for variations and TrueType hinting:
+   we detect contiguous sequences of unmodified points and interpolate them
+   based on the adjustments made to the points on the boundary.
+
+### Grid-fitting complete
+
+And with that, we have a nicely hinted outline! Simple.
```

Comment:
Suggest adding a references block for http://rastertragedy.com/ and the link Behdad mentioned.